src/branch_and_bound.rs

```diff
@@ -0,0 +1,582 @@
+        //        2
+        //       /
+        //      1
+        //
+        // To:
+        //      o
+        //     / \
+        //        4
+        //       / \
+        //      3
+        //
+        //
+        // 2_ value meets or exceeded target.
+        //    In this condition, we only backtrack one node
+        //
+        // From:
+        //      o
+        //     /
+        //    4
+        //   /
+        //  3
+        //
+        // To:
+        //      o
+        //     /
+        //    4
+        //   / \
+        //      3
+
+        // Set initial loop state
+        backtrack = false;
+        backtrack_subtree = false;
+
+        // * not enough value to make it to the target.
+        //   Therefore, explore a new new subtree.
+        if available_value + value < target {
+            backtrack_subtree = true;
+        }
+        // This optimization provides an upper bound on the amount of waste that is acceptable.
+        // Since value is lost when we create a change output due to increasing the size of the
+        // transaction by an output, we accept solutions that may be larger than the target as
+        // if they are exactly equal to the target and consider the overage waste or a throw
+        // away amount.  However we do not consider values greater than value + cost_of_change.
+        //
+        // This effectively creates a range of possible solution where;
+        // range = (target, target + cost_of_change]
+        //
+        // That is, the range includes solutions that exactly equal the target up to but not
+        // including values greater than target + cost_of_change.
+        else if value > target + cost_of_change {
+            backtrack = true;
+        }
+        // * value meets or exceeds the target.
+        //   Record the solution and the waste then continue.
+        //
+        // Check if index_selection is better than the previous known best, and
+        // update best_selection accordingly.
+        else if value >= target {
+            backtrack = true;
+
+            let current_waste = value - target;
+
+            if current_waste <= waste {
+                best_selection = Some(index_selection.clone());
+                waste = current_waste;
+            }
+        }
+
+        // * Backtrack
+        if backtrack {
+            let last_index = index_selection.pop().unwrap();
+            value -= utxo_candidate_amounts[last_index];
+            index -= 1;
+            assert_eq!(index, last_index);
+        }
+        // * Backtrack to new tree
+        else if backtrack_subtree {
+            // No new subtree left to explore.
+            if index_selection.is_empty() {
+                return index_to_utxo_list(best_selection, weighted_utxos);
+            }
+
+            // Anchor the new subtree at the next available index.
+            // The next iteration, the index will be incremented by one.
+            index = index_selection.remove(0);
```

Comment:
This requires shifting the whole thing, can the algorithm be rewritten to use `VecDequeue`?

Reply:
Not sure what you mean by shifting the whole thing.  Although depending on where the pointer is that rust maintains to the vec, maybe it has to do the arithmetic to find the head and remove?  VecDeque should be faster if pop_front() can be used since there's no pointer arithmetic involved.  Will test that idea out.  thanks!

Reply:
When you call `remove(0)` all contents of `Vec` has to be copied to take its place. Pointer arithmetic is actually infinitely faster and it's what `VecDequeue` uses. `pop_front` is equivalent to `remove(0)` in semantics but much faster.

Reply:
Ah I see, yeah copying the entire vec where the 2nd becomes the first etc would be slow.

Reply:
I tested out a DeQueue in place of a Vec and surprisingly it's far slower.  Went from about 830,000 to 950,000 or so.  Here is the experimental commit:

https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/28/commits/18145ce0c61f1da6d191023f4970a4eb727ad1e7

Reply:
Looking at Core code I don't see remove there. Maybe if this code resembled it more it'd be faster? I don't understand the code enough to understand the difference (so far I only scanned obvious perf issues).

Reply:
Do I understand correctly that the Core code ignores the first value rather than popping it? So it keeps it in memory? If so, I think that's what makes it faster.

Reply:
No, it does pop the value off `curr_selection` using `pop_back()` https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.cpp#L147.   So in the case where the target is 5, and the solution index [0, 3] was found, the for loop unwinds and index and 3 is popped, then the next iteration occurs, and backtrack happens again, except it unwinds the index back to 0 which is then popped, and the next iteration starts at 1.  So essentially, `pop_back()` is equivalent to the what I've done with `remove(0)`,  I just do it with less iterations.

Reply:
Actually, come to think of it, we can just do a memory access here instead of `pop()`.  I tested the change and it reduces the time from 830k ns/iter to 810k ns/iter.  The reason is, we are just doing a clear later.

Reply:
Popping from back is completely different - and much faster! I really don't understand the code yet (and am very tired right now) so are you doing everything in reverse?

Reply:
> Popping from back is completely different - and much faster! I really don't understand the code yet (and am very tired right now) so are you doing everything in reverse?

Yes, however in the case where we are at the last index, the following occurs:

Returning to the example of `[0, 3]`.  At this point, the C++ code does a `pop_back()` so the selection becomes `[0]` then another iteration occurs.  Next iteration, it now does `pop_back()` on `[0]` so we are at an empty list.  This actually isn't very efficient for the edge case where the C++ code does multiple pop_back() functions to backtrack to a new subtree.  An equivalent way to represent this edge case is to store `0` and clear the list instead of 2 `pop_back()` calls.  Finally the other case where we are back tracking one node, I also do a pop back here: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/28/files#diff-9098d62be93e83524a8371395c973d761a95000d1c295f600a8c808e917c16d9R24

Reply:
Well, if you're calling `remove(0)` then you also iterate twice, you just don't see it because the method does it.

Reply:
This has been changed to `remove[0]`: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/28/files#diff-9098d62be93e83524a8371395c973d761a95000d1c295f600a8c808e917c16d9R254

Reply:
Also, I think you're confusing which iteration I'm talking about.  The way core is doing it, when a backtrack occurs that must back track all the way to the root, it's doing two iterations of the BnB loop.

Reply:
Yeah, I still didn't have time to study the algo...

Reply:
Understandable.  It takes a bit of study to really understand how the algo works.  I was thinking it would valuable to benchmark the number of utxo combinations that are tried per time interval instead of the number of the time per iteration since sometimes a full iteration is spent just backtracking one or more steps.  So to get a real comparison of how well the algos perform, some better code profiling should be helpful.

Reply:
I guess smarter people than us already figured that out and just correctly porting the code is the most valuable thing to do right now.

Reply:
Yeah it would be nice to have some people who know that part of bitcoin core review this.  I could try an exact copy of the method core uses to backtrack, maybe on a different branch.  Would be interesting to see if there's a difference in timing.